Client/NoOpRunner.Client/Rendering/GameWindowRenderer.cs

```diff
@@ -1,48 +1,188 @@
+
+                    Canvas.SetLeft(canvas.Children[i], rectangleWidth * pixels[i].X);
+                    Canvas.SetBottom(canvas.Children[i], rectangleHeight * pixels[i].Y);
+                }
+            }
 
-                Canvas.SetLeft(rec, rectangleWidth * pixel.X);
-                Canvas.SetBottom(rec, rectangleHeight * pixel.Y);
+            #region Old code
 
-                rec.MouseLeftButtonDown += (s, e) =>
+// this still retarded, why generate empty pixels?? make canvas background color
+            // var pixels = window.GetCurrentWindow();
+            // canvas.Children.Clear();
+            // foreach (var pixel in pixels)
+            // {
+            //     
+            //     var rec = new Rectangle
+            //     {
+            //         Width = rectangleWidth,
+            //         Height = rectangleHeight,
+            //         Fill = ColorBrushMap[pixel.Color]
+            //     };
+            //
+            //     Canvas.SetLeft(rec, rectangleWidth * pixel.X);
+            //     Canvas.SetBottom(rec, rectangleHeight * pixel.Y);
+            //
+            //
+            //     rec.MouseLeftButtonDown += (s, e) => { window.ClickShape(pixel.X, pixel.Y); };
+            //
+            //     canvas.Children.Add(rec);
+            // }
+
+            #endregion
+        }
+
+        public static void RenderPlayer(Player player, Canvas canvas, GameWindow window)
+        {
+            //TODO: player size, adjust animation speed for better UX 
+            var rectangleWidth = canvas.ActualWidth / window.SizeX;
+            var rectangleHeight = canvas.ActualHeight / window.SizeY;
+
+            var playerPixels = player.Render();
+
+            if (canvas.Children.Count == playerPixels.Count)
+            {
+                for (int i = 0; i < playerPixels.Count; i++)
                 {
-                    window.ClickShape(pixel.X, pixel.Y);
-                };
+                    
+                    if (player.StateMachine.CompareStates)
+                    {
+                        canvas.Children[i].SetValue(GifImage.GifSourceProperty,
+                            player.StateMachine.GetStatusUri());
+                        
+                        canvas.Children[i].SetValue(Canvas.WidthProperty, rectangleWidth);
+                        canvas.Children[i].SetValue(Canvas.HeightProperty, rectangleHeight);
+                    }
 
-                canvas.Children.Add(rec);
+                    if (player.StateMachine.IsTurning)
+                    {
+                        canvas.Children[i].SetValue(UIElement.RenderTransformProperty,
+                            player.StateMachine.IsTurnedLeft
+                                ? new ScaleTransform() {ScaleX = -1}
+                                : new ScaleTransform() {ScaleX = 1});
+                    }
+
+                    #region Animation rectangle move, need non static method
```

Comment:
Why is this region tag here? Does it really make a meaningful impact to put this one statement in a separate region? If you're using regions, then use them everywhere or nowhere. This could have just been a comment.

Reply:
Yes that is from past, now we move on